src/Renderer/Simulator/CanvasStateAnalyser.fs

```diff
@@ -471,20 +483,63 @@ let checkCustomComponentsOk ((comps, _): CanvasState) (sheets: LoadedComponent l
         | Error(c, BadOutputs(cName, instOuts, compOuts)) ->
             let instOus, compOuts = disp instOuts, disp compOuts
 
-            error c
+            error c OutPortMismatch
             <| sprintf
                 "Sheet %s is used as a custom component. Instance Out ports: %A are different from Component Out ports: %A."
                 cName
                 instOuts
                 compOuts
 
+let getPortNumFromConnPort (connPort: Port) (targetPortList: Port list) =
+    targetPortList
+    |> List.tryFind (fun (port: Port) -> port.Id = connPort.Id)
+    |> function
+        | Some {PortNumber = Some n} -> n
+        | _ -> failwithf "Existing connection should be connected to existing port"
+
+/// Check whether Adders have a NotConnected component connected to their COUT
+/// this is unnecessary since it can be disabled via its options
+let checkAdderUnnecessaryNC ((comps, conns): CanvasState) : SimulationError option =
+    let isIdofNotConnectedComp id =
+        comps
+        |> List.exists (fun comp ->
+            comp.Id = id && comp.Type = NotConnected)
+    
+    let compLFold (affComps, affConns) (comp: Component) =
```

Comment:
Names here are v obscure. Using Fold here should be as last resort. I think the whole of checkAdderUnnecessaryNC could be done simpler without fold list functions. Maybe you need to return a "list of unconnected ports" to avoid later repetition in which case the filter is in fact just a List..concat - and tyehresult is [] if nothing or teh lits of unconnected ports.

Lists and lits fucntions are your friend. use them. Over-use of Fold is juts you thinking imperatively.

condition =
match
| (if Adder) ->
     output ports
     |> select not connected

components
-> filter (condition)